src/core/lib/slice/slice.cc

```diff
@@ -257,94 +208,44 @@ grpc_slice grpc_slice_from_cpp_string(std::string str) {
-
-  grpc_slice_refcount* base_refcount() { return &base_; }
-
- private:
-  grpc_slice_refcount base_;
-  std::atomic<size_t> refs_{1};
-};
-
-}  // namespace
-
 grpc_slice grpc_slice_malloc_large(size_t length) {
-  return grpc_core::UnmanagedMemorySlice(
-      length, grpc_core::UnmanagedMemorySlice::ForceHeapAllocation());
-}
-
-void grpc_core::UnmanagedMemorySlice::HeapInit(size_t length) {
-  /* Memory layout used by the slice created here:
-
-     +-----------+----------------------------------------------------------+
-     | refcount  | bytes                                                    |
-     +-----------+----------------------------------------------------------+
-
-     refcount is a malloc_refcount
-     bytes is an array of bytes of the requested length
-     Both parts are placed in the same allocation returned from gpr_malloc */
-  auto* rc =
-      static_cast<MallocRefCount*>(gpr_malloc(sizeof(MallocRefCount) + length));
-
-  /* Initial refcount on rc is 1 - and it's up to the caller to release
-     this reference. */
-  new (rc) MallocRefCount();
-
-  /* Build up the slice to be returned. */
-  /* The slices refcount points back to the allocated block. */
-  refcount = rc->base_refcount();
-  /* The data bytes are placed immediately after the refcount struct */
-  data.refcounted.bytes = reinterpret_cast<uint8_t*>(rc + 1);
-  /* And the length of the block is set to the requested length */
-  data.refcounted.length = length;
+  grpc_slice slice;
+  uint8_t* memory = new uint8_t[sizeof(grpc_slice_refcount) + length];
+  slice.refcount = new (memory) grpc_slice_refcount(
+      [](grpc_slice_refcount* p) { delete[] reinterpret_cast<uint8_t*>(p); });
+  slice.data.refcounted.bytes = memory + sizeof(grpc_slice_refcount);
+  slice.data.refcounted.length = length;
+  return slice;
 }
 
 grpc_slice grpc_slice_malloc(size_t length) {
-  return grpc_core::UnmanagedMemorySlice(length);
-}
-
-grpc_core::UnmanagedMemorySlice::UnmanagedMemorySlice(size_t length) {
-  if (length > sizeof(data.inlined.bytes)) {
-    HeapInit(length);
+  if (length <= GRPC_SLICE_INLINED_SIZE) {
+    grpc_slice slice;
+    slice.refcount = nullptr;
+    slice.data.inlined.length = length;
+    return slice;
   } else {
-    /* small slice: just inline the data */
-    refcount = nullptr;
-    data.inlined.length = static_cast<uint8_t>(length);
+    return grpc_slice_malloc_large(length);
   }
 }
 
-template <typename Slice>
-static Slice sub_no_ref(const Slice& source, size_t begin, size_t end) {
-  Slice subset;
+static grpc_slice sub_no_ref(const grpc_slice& source, size_t begin,
+                             size_t end) {
+  grpc_slice subset;
 
   GPR_ASSERT(end >= begin);
 
-  if (source.refcount) {
+  if (source.refcount != nullptr) {
     /* Enforce preconditions */
     GPR_ASSERT(source.data.refcounted.length >= end);
 
     /* Build the result */
-    subset.refcount = source.refcount->sub_refcount();
+    subset.refcount = source.refcount;
```

Comment:
Not a new thing, but it seems a bit odd to propagate the ref-count, when the point of this function is to return a sub-slice that does *not* take a ref.  I'm guessing that this is necessary for operations on the new sub-slice that may need to check the type of ref-counting done on the parent slice?

Reply:
That and you might want to later ref it.